src/playmp_controller.cpp

```diff
@@ -270,32 +270,20 @@ void playmp_controller::wait_for_upload()
 	// If the host is here we'll never leave since we wait for the host to
 	// upload the next scenario.
 	assert(!is_host());
+	// TODO: should we handle the case that we become the ho0st because the host disconnectes here?a
 
-	config cfg;
-	network_reader_.set_source(playturn_network_adapter::get_source_from_config(cfg));
-
-	while(true) {
-		try {
-			bool res = false;
-			gui2::dialogs::loading_screen::display([&]() {
-				gui2::dialogs::loading_screen::progress(loading_stage::next_scenario);
+	gui2::dialogs::loading_screen::display([&]() {
+		gui2::dialogs::loading_screen::progress(loading_stage::next_scenario);
 
-				res = mp_info_->connection.wait_and_receive_data(cfg);
-			});
-
-			if(res && turn_data_.process_network_data_from_reader() == turn_info::PROCESS_END_LINGER) {
-				break;
-			} else {
-				throw_quit_game_exception();
-			}
-		} catch(const quit_game_exception&) {
-			network_reader_.set_source([this](config& cfg) { return receive_from_wesnothd(cfg); });
-			turn_data_.send_data();
-			throw;
+		auto res = turn_data_.process_network_data_from_reader();
+		if(res == turn_info::PROCESS_END_LINGER) {
+			return;
+		} else if (res != turn_info::PROCESS_CONTINUE) {
+			throw quit_game_exception();
 		}
-	}
-
-	network_reader_.set_source([this](config& cfg) { return receive_from_wesnothd(cfg); });
```

Comment:
well, it should, the first set_source changes the "source" and then the second two reset it to the original source. but both sources call `mp_info_->connection.wait_and_receive_data`in the end.

Reply:
Even if it should work, it was tested, right?

Reply:
This one probably not, since it seemed pretty easy but i can do that, sure.

Reply:
Please don't ever push untested code into master.

Reply:
Hmm seems like i missed putting the while loop there, it was clearly there in my mind :/

But whats more important: it was already missing a `turn_data_.send_data();` call here before this commit to send the last action, i have no idea why the mp tests passed before this.

Reply:
tbh the whole "When do we send actions" should maybe be refactored or at least investigated. But that not within the scope of this pr since its kinda complicated in particular because we have to make sure we don't send undoable actions too early.